packages/react-server/src/ReactFlightServer.js

```diff
@@ -698,6 +699,8 @@ function renderElement(
     // When the ref moves to the regular props object this will implicitly
     // throw for functions. We could probably relax it to a DEV warning for other
     // cases.
+    // TODO: `ref` is now just a prop when `enableRefAsProp` is on. Should we
+    // do what the above comment says?
```

Comment:
We should really disable the error here when the flag is on and instead error inside the host component path instead, where it can check `props.ref` when the flag is on. Class components already error anyway.

When a "ref" prop is passed with a function to a client component it'll just be a general error but we could special case that message similar to how special cased function passed to children here:

https://github.com/facebook/react/pull/28367